[Work in progress] Add isolation metric

khl02007 opened the pull request:
This PR adds a new metric called isolation metric to `toolkit.qualitymetrics`. It is described in [Chung 2017](https://doi.org/10.1016/j.neuron.2017.08.030). 
* There is already an implementation of this metric in the form of `nn_hit_rate` in `nearest_neighbors_metrics` function in `misc_metrics.py`. However, this implementation has a couple of issues:
  * It fails to handle the case in which `n_neighbors`=1 (returns nan).
  * It is also _not the same_ as the isolation metric from Chung 2017, even though the docstring claims that it is. This difference is known and is discussed [here](https://github.com/SpikeInterface/spikemetrics#a-note-on-calculations).
* While it is true that the metric can be unreliable when comparing two clusters with a large difference in firing rate, Chung 2017 actually addresses this point and suggests randomly subsampling the two clusters such that the number of spikes in each cluster is equal before computing isolation. 
* The `nearest_neighbors_isolation` function in this PR implements the random subsampling as described in Chung 2017 with a seed parameter to set a pseudorandom number generator. 
* Right now the new function is not referenced by others (e.g. `calculate_pc_metrics`). @alejoe91 Once you approve adding this metric I will work on these.

alejoe91 commented:
@khl02007 thank you so much for this PR and thanks for solving the issues you mentioned. :)
The implementation looks good to me, so please go ahead and add it to the `calculate_pc_metrics`!

khl02007 commented:
@alejoe91 
* changed name to `nearest_neighbor_isolation` (`nn_isolation`)
* added another argument (`seed`) to `calculate_pc_metrics`
* added code to handle kwargs for pca metrics in `QualityMetricCalculator.compute_metrics`
* what to add in `quality_metric_list.py`?

alejoe91 commented:
@khl02007 I'm currently on vacation until next week. I think we still need to incorporate the pc metrics in the full metrics list, but let's discuss about it next week :)

samuelgarcia commented:
Hi @khl02007.
Thank you very much for this.
I will have a deep look on this very soon.

I ported all metrics very quickly into this new spiekinetrface API.
They all need to be check carfully.

Note that the metrics are computed on top on WaveformExtarctor that already sub sample the spike count for each cluster.

samuelgarcia commented on spikeinterface/toolkit/qualitymetrics/quality_metric_calculator.py:
```diff
@@ -38,11 +38,31 @@ def compute_metrics(self, metric_names=None, **kwargs):
+        if 'seed' in kwargs:
+            seed = kwargs['seed']
+        else:
+            seed = 0
```
I really would prefer seed=None as default seed.

khl02007 replied:
Thanks, changed.

khl02007 commented:
@samuelgarcia No worries, this was not your mistake - the nearest neighborhood metrics were implemented differently from Chung 2017 in the older `spikemetrics` package as well. This is not necessarily incorrect (see the discussion in the README of `spikemetrics`), which is why I did not propose getting rid of `nn_hit_rate` or `nn_miss_rate` but simply to add a new metric that more faithfully corresponds to what is described in Chung 2017 (and what our lab uses).

lfrank commented:
Hi all. Just checking in on this; we would love to be able to use the isolation metric for our sorts. Does someone have time to review this?

alejoe91 commented:
Hi Loren, this week both Samuel and I are busy with some summer schools. We'll review the code and coordinate with @khl02007 to finish this up next week. Also please feel free to make a PR also to SI-0.13 (spiketoolkit). I can make a new release once it's merged :)

samuelgarcia commented on spikeinterface/toolkit/qualitymetrics/pca_metrics.py:
```diff
@@ -11,7 +11,7 @@
 _possible_pc_metric_names = ['isolation_distance', 'l_ratio', 'd_prime', 'nearest_neighbor']
```
You need to add 'nn_isolation' here.
So it is tested and integrated magically in the table.

samuelgarcia commented:
This looks to me.
@alejoe91 : I let you have a look also.

alejoe91 commented:
@samuelgarcia @khl02007 this looks good to me as well :)

lfrank commented:
Thanks much…

> On Sep 6, 2021, at 1:29 AM, Garcia Samuel ***@***.***> wrote:
> 
> ﻿
> Merged #220 into master.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.
